packages/material-ui/src/NoSsr/NoSsr.js

```diff
@@ -11,17 +10,15 @@ import { exactProp } from '@material-ui/utils';
  * - Reduce the rendering time on the server.
  * - Under too heavy server load, you can turn on service degradation.
  */
-class NoSsr extends React.Component {
-  mounted = false;
+function NoSsr(props) {
+  const { children, defer, fallback } = props;
+  const mountedRef = React.useRef(false);
+  const [mountedState, setMountedState] = React.useState(false);
 
-  state = {
-    mounted: false,
-  };
+  React.useEffect(() => {
```

Comment:
I would change the approach. There are different cases to consider. Basically, the key objective is to limit UI jumps with NoSsr. You can't do anything about it on the server, but you can do on the client.
I'm proposing:
```jsx

const baseEffect = typeof window === 'undefined' ? React.useEffect : React.useLayoutEffect;

// This variable will be true once the server-side hydration is completed.
let hydrationCompleted = false;

function NoSsr(props) {
  const { children, defer, fallback } = props;
  const [mountedState, setMountedState] = React.useState(() => {
    if (hydrationCompleted && !defer) {
       return true;
    }

    return false;
  });

  baseEffect(() => {
    if (!defer) {
      setMountedState(true);
    }
  }, [defer]);

  React.useEffect(() => {
    hydrationCompleted = true;

    if (defer) {
      setMountedState(true);
    }
  }, [defer]);

  return <React.Fragment>{mountedState ? children : fallback}</React.Fragment>;
}
```

You can play with this demo:
```jsx
import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from '@material-ui/core/styles';
import NoSsr from '@material-ui/core/NoSsr';
import Button from '@material-ui/core/Button';

const styles = theme => ({
  button: {
    display: 'block',
    margin: theme.spacing(2),
  },
});

function LargeTree() {
  return Array.from(new Array(1000)).map((_, index) => <button key={index}>{index}</button>);
}

// to play with
const defer = true;

function SimpleNoSsr(props) {
  const { classes } = props;
  const [open, setOpen] = React.useState(false);

  return (
    <div>
      <button onClick={() => setOpen(!open)}>Hello</button>
      {open ? (
        <div>
          <span>butu</span>
          <NoSsr defer={defer}>
            <LargeTree />
          </NoSsr>
        </div>
      ) : null}
    </div>
  );
}

SimpleNoSsr.propTypes = {
  classes: PropTypes.object.isRequired,
};

export default withStyles(styles)(SimpleNoSsr);
```

Reply:
We can keep the parts of the previous comment that make sense.

Reply:
> Basically, the key objective is to limit UI jumps with NoSsr. 

Wouldn't this still be the case for your example on the first hydration?

Why not just `return isInBrowser ? children : null;`

Reply:
> Why not just return isMounted ? children : null; where isMounted is set during an effect

I originally thought this

Reply:
@eps1lon You would need to rerender.

Reply:
Isn't that the point? I thought that was better than changing the server markup during the first commit.

Reply:
Could you provide a more detailed implementation? I fail to capture the global solution.

Reply:
I won't have time to look into this. The [part in the docs ](https://reactjs.org/docs/hooks-reference.html#uselayouteffect) is specifically talking about hiding components that use layout effect on the server. I would suspect that this is a use case of `NoSsr`. 

Could you open an issue in react verifying that your implementation is better for that use case?

Reply:
Yes, it's exactly what my implementation proposal uses.

Reply:
You're using a layout effect to display a children. React recommends a normal effect.

Reply:
It's what the defer property is for. This pull request is about migrating from the classes API to hooks. Using useEffect is a breaking change in the behavior.

Reply:
Fair enough. `defer` would work different with your implementation as well though so that would be a breaking change as well. We should should probably add a use case for `defer` explaining that this should be used to hide components that need layout effects.

Reply:
The way I see it, defer is not for hiding a component that needs layout effects. The defer property is to control how React commit the react tree. Does it do it in 2 or 1 chunk.

Reply:
With defer=true, it takes 2 commits. With defer=false it's a single commit.

Reply:
defer has no influance on hydration.

Reply:
> With defer=true, it takes 2 commits. With defer=false it's a single commit.

I wouldn't be so sure about this with the current implementation and concurrent mode. I think all the `setImmediate` and `rAF` calls are if not unsafe at least very hard to reason about in concurrent mode.

Reply:
I have never tried concurrent mode. Have you, or can we?
We could add a demo for the defer prop so people understand who it behaves. Maybe React will better address the problem in the core. It's quite a moving field! I would vote for keeping the same behavior (with my proposed logic, not using rAF directly), and see how it evolves.

Reply:
> I have never tried concurrent mode. Have you, or can we?

Yes but only played with it a bit. You can import it as `unstable_ConcurrentMode` but I don't think you should derive anything from the behavior. I'm a little bit more confident now when we can accept side-effects in render and when not but I still try to avoid these things as best as I can because any bug they will cause will almost certainly be non-reproducible i.e. Heisenbug.

Maybe `setImmediate` or `rAF` is totally fine in ConcurrentMode but it would be better to give the control to react (or rather the scheduler) to apply some better heuristics.